azurerm/resource_arm_log_analytics_solution.go

```diff
@@ -0,0 +1,214 @@
+	parameters := operationsmanagement.Solution{
+		Name:     &name,
+		Location: &location,
+		Plan:     &solutionPlan,
+		Properties: &operationsmanagement.SolutionProperties{
+			WorkspaceResourceID: &workspaceID,
+		},
+	}
+
+	res, err := client.CreateOrUpdate(ctx, resGroup, name, parameters)
+	//Currently this is required to work around successful creation resulting in an error
+	// being returned
+	if err != nil && res.StatusCode != 201 {
+		return err
+	}
+
+	solution, _ := client.Get(ctx, resGroup, name)
+
+	if solution.ID == nil {
+		return fmt.Errorf("Cannot read Log Analytics Solution '%s' (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*solution.ID)
+
+	return resourceArmLogAnalyticsSolutionRead(d, meta)
+
+}
+
+func resourceArmLogAnalyticsSolutionRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).solutionsClient
+	ctx := meta.(*ArmClient).StopContext
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["solutions"]
+
+	resp, err := client.Get(ctx, resGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on AzureRM Log Analytics solutions '%s': %+v", name, err)
+	}
+
+	if resp.Plan == nil {
+		return fmt.Errorf("Error making Read request on AzureRM Log Analytics solutions '%s': %+v Plan was nil", name, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("location", resp.Location)
+	d.Set("resource_group_name", resGroup)
+	d.Set("plan", flattenAzureRmLogAnalyticsSolutionPlan(*resp.Plan))
+	return nil
+}
+
+func resourceArmLogAnalyticsSolutionDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).solutionsClient
+	ctx := meta.(*ArmClient).StopContext
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["solutions"]
+
+	resp, err := client.Delete(ctx, resGroup, name)
+
+	if err != nil {
+		if utils.ResponseWasNotFound(resp) {
+			return nil
+		}
+
+		return fmt.Errorf("Error issuing AzureRM delete request for Log Analytics Solution '%s': %+v", name, err)
+	}
+
+	return nil
+}
+
+func expandAzureRmLogAnalyticsSolutionPlan(d *schema.ResourceData) operationsmanagement.SolutionPlan {
+	plans := d.Get("plan").([]interface{})
+	plan := plans[0].(map[string]interface{})
+
+	expandedPlan := operationsmanagement.SolutionPlan{}
+
+	if name := plan["name"].(string); len(name) > 0 {
+		expandedPlan.Name = &name
+	}
+
+	if publisher := plan["publisher"].(string); len(publisher) > 0 {
+		expandedPlan.Publisher = &publisher
+	}
+
+	if promotionCode := plan["promotion_code"].(string); len(promotionCode) > 0 {
+		expandedPlan.PromotionCode = &promotionCode
+	} else {
+		blankString := ""
+		expandedPlan.PromotionCode = &blankString
+	}
+
+	if product := plan["product"].(string); len(product) > 0 {
+		expandedPlan.Product = &product
+	}
```

Comment:
Looks much cleaner, I'll update - I was being over cautious as new to the terraform schema idea. 

Only think I'll have to keep is the check for an empty `promotion_code` as the API returns a `400` if the field is omitted :( although it is optional. I need to ensure it's always at least an empty string rather than nil otherwise go will drop the field when marshalling.